### PR TITLE
Additional Device Compiler Tests

### PR DESCRIPTION
**Pull request by agozillon**
Several new device compiler tests that work with POCL when they're
compiled using a newer version of the triSYCL device compiler (after the
commit: [OpenCL] Add generic AS to 'this' pointer).

However, the buffer_of_objects_members.cpp test works with the
sycl/release_70/master branch of the triSYCL device compiler.

The buffer of objects tests are related to:
https://github.com/triSYCL/triSYCL/issues/212

The ctor_and_dtor tests are related to:
https://github.com/triSYCL/triSYCL/issues/213

### CODE REVIEW

**Comment by agozillon** [comment]
If there are any other useful things I can test and add to this commit that are related to issue 213 and 212 I'd be happy to add them

File: tests/device_compiler/parallel_for_class_ctor_and_dtor.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+  my_queue.submit([&](handler &cgh) {
+    auto acc = a.get_access<access::mode::write>(cgh);
+
+    Index_Member_Data idm_inner;
+    idm_inner.set_g(10);
+    idm_inner.set_r(10);
+
+    cgh.parallel_for<class ctor_dtor>(
+        range<1>{N}, [&, d_a = drt::accessor<decltype(acc)>{acc}](item<1> index) {
```
Comment (keryell) [inline tests/device_compiler/parallel_for_class_ctor_and_dtor.cpp]: I do not think you can use `&` on a device where there is no unified memory with the host.
Furthermore, even on CPU, the kernel is executed in another thread and `idm_inner` is probably destructed already when the kernel start.
This is why in SYCL we have only capture by value.

File: tests/device_compiler/single_task_class_ctor_and_dtor.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    idm_outer.set_g(20);
+
+    my_queue.submit([&](handler &cgh) {
+      auto acc = a.get_access<access::mode::write>(cgh);
+
+      Index_Member_Data idm_inner;
+      idm_inner.set_g(10);
+
+      cgh.single_task<class ctor_dtor>(
+        [&, d_a = drt::accessor<decltype(acc)>{acc}] {
```
Comment (keryell) [inline tests/device_compiler/single_task_class_ctor_and_dtor.cpp]: Same problem. `=` instead of `&`

**Comment by agozillon** [comment]
Just an update, I tested with xocc for both tests on hw_emu and sw_emu, it works for both hw_emu and sw_emu for the buffer_of_object_members.cpp test.

However, it only works for sw_emu on the buffer_of_objects_member_functions.cpp test, it crashes during compilation when compiling from kernel.xpirbc to kernel.xo using the xocc compiler. 

The input xpirbc file seems to be the same in both sw_emu and hw_emu (as I believe it's just the llvm bc file with a name swap and preserved metadata). So perhaps hw_emu compilation does an extra step and finds fault with the input.

It's unfortunately a little difficult to debug right now as only xocc appears to be in debug build and not the clang compiler it invokes, it does throw an assertion at a specific line of llvm (although, I don't seem to be able to access the source as it's in a directory I have no access to at the moment) so I have some idea to go off of in this case: 

llvm::Value::~Value(): Assertion `use_empty() && "Uses remain when a value is destroyed!"' failed. 

xocc also outputs a more detailed handy clang output log that states:

While deleting: void %
Use still stuck around after Def is destroyed:  %4 = addrspacecast void <badref> to %struct.Index_Member_Data*

Although the IR seems to be a little different than the input .xpirbc file at this point, but it seems to be incorrectly cleaning away uses of a value it tried to optimize out. Still looking into the problem though, so it could be something unrelated to that assumption!

**Comment by keryell** [comment]
You need to ask Alexandre about how to use a debug build of the xocc compiler itself.

**Comment by keryell** [comment]
It looks like we should not generate these address space cast from `addrspace(1)*`.
This comes from the fact that all the member functions operate on address space 0 for now.
It could be solved at the Clang codegen level like @AnastasiaStulova & co are doing to improve address space support in C++ or in a simple way by adding a SYCL LLVM pass to do some trivial address space inference, such as propagating the address spaces from `addrspacecast .. to` instruction.
In the case above, `%0` should remain in `addrspace(1)` and we should propagate this address space downward by using some fix point computation for example.

**Comment by keryell** [comment]
Thanks for the new tests!

**Pull request merged**